### ISSUE TITLE
sustain pedal not being recorded in piano roll

### ISSUE THREAD

**Issue by durango99** [LMMS/lmms#1700]
Windows 8.1 64bit
LMMS 1.1.0 64bit (and 32bit)

I've seen 2 other issues noted that were fixed earlier last year on this but I'm not able to record the sustain.  I assume my issue is the same and the MIDI data sent by my midi keyboard is not something LMMS recognizes.

I've logged the midi here:

 TIMESTAMP IN PORT STATUS DATA1 DATA2 CHAN NOTE EVENT  
 0005C341   2   2     B0    40    7F    1  ---  CC: Pedal (Sustain)  
 0005C436   2   2     B0    40    00    1  ---  CC: Pedal (Sustain)   

Thanks for any help you can provide.

**Comment by tresf** [comment]
Thanks.  We'll take a look. 

Related: #168 (Closed)

**Comment by gligeti** [comment]
Issue still present in 1.1.3. Pedal events with values (00-7F) as above are not recorder. Ideally these should be recorded as part as the automation recording.
Makes LMMS useless for recording serious piano/keyboard tracks.

**Comment by oeai** [comment]
still not fixed

**Comment by tresf** [comment]
> still not fixed

@oeai we have 439 bugs that this statement is true for.  We need help to fix these, it is greatly appreciated.

**Comment by oeai** [comment]
@tresf i can say that in fact sustain is working, but just not recording
maybe it's a part of missing automation? because in piano-roll i suppose there's no such event if we only record notes and there should be some more complex behaviour maybe?

**Comment by plumpNation** [comment]
Is there a workaround for this issue? I would love to help but my coding skills simply aren't up to the job.

**Comment by plumpNation** [comment]
Is this fixed in 1.2?

**Comment by zonkmachine** [comment]
No, this is still an issue. I'm looking into it right now.

**Comment by zonkmachine** [comment]
The bug was introduced in 0.9.91/ b28b00ff2346f81694d821d39736fb6b58cc089b , one month after recording with sustain was fixed previously. The actually note recorded is defined by the quantization setting.

```
b28b00ff2346f81694d821d39736fb6b58cc089b is the first bad commit
commit b28b00ff2346f81694d821d39736fb6b58cc089b
Author: Tobias Doerffel <tobias.doerffel@gmail.com>
Date:   Mon Feb 24 23:24:11 2014 +0100

    NotePlayHandle: emit midiNoteOff() signal in noteOff()
    
    Instead of emitting InstrumentTrack::midiNoteOff() in destructor of
    NotePlayHandle do this where it actually happens -> noteOff().
    
    Fixes length of recorded notes when there's e.g. a long release.
    
    Closes #378.
```

**Comment by zonkmachine** [comment]
If you activate the 'Stacking' function it will work under sustained notes but the 'Arpeggiator' won't. I don't know if there is a good reason for this but there probably isn't. This is true in 1.1.3 as well.

**Comment by serdnab** [comment]
@zonkmachine  are you still on this?
I'm interested on fixing it.

**Comment by zonkmachine** [comment]
> I'm interested on fixing it.

Code away! :+1:

**Comment by zonkmachine** [comment]
@serdnab Did you look into this?

**Comment by serdnab** [comment]
@zonkmachine 
I'm currently working on it.

**Comment by zonkmachine** [comment]
Cool. I'll test anything you throw at us! ;)

**Comment by zonkmachine** [comment]
Fixed in https://github.com/LMMS/lmms/pull/3710
There is one more bug however to fix before the sustain pedal will be fully functional.
#3537

**Comment by zonkmachine** [comment]
> There is one more bug however to fix before the sustain pedal will be fully functional.
> #3537

Which was closed in https://github.com/LMMS/lmms/pull/3730, again by @serdnab 
The sustain pedal now works and the changes will be included in LMMS-1.2.0-RC4 which will be out, soonish.
